kentikapi/client.go

```diff
@@ -0,0 +1,133 @@
+)
+
+// Client is Kentik API client.
+type Client struct {
+	config     Config
+	httpClient *http.Client
+}
+
+func NewClient(c Config) *Client {
+	if c.APIURL == "" {
+		c.APIURL = APIURLUS
+	}
+
+	return &Client{
+		config: c,
+		httpClient: &http.Client{
+			Transport: &http.Transport{
+				Proxy: http.ProxyFromEnvironment,
+				DialContext: (&net.Dialer{
+					Timeout:   30 * time.Second,
+					KeepAlive: 30 * time.Second,
+				}).DialContext,
+				ForceAttemptHTTP2:     true,
+				MaxIdleConns:          100,
+				IdleConnTimeout:       90 * time.Second,
+				TLSHandshakeTimeout:   10 * time.Second,
+				ExpectContinueTimeout: 1 * time.Second,
+			},
+		},
+	}
+}
+
+type Config struct {
+	// APIURL defaults to "https://api.kentik.com/api/v5"
+	APIURL    string
+	AuthEmail string
+	AuthToken string
+}
+
+const (
+	authEmailKey    = "X-CH-Auth-Email"
+	authAPITokenKey = "X-CH-Auth-API-Token"
+)
+
+func (c *Client) GetAllUsers(ctx context.Context) (_ []User, err error) {
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, c.config.APIURL+"/users", nil)
+	if err != nil {
+		return nil, fmt.Errorf("new request: %v", err)
+	}
+
+	request.Header.Set(authEmailKey, c.config.AuthEmail)
+	request.Header.Set(authAPITokenKey, c.config.AuthToken)
+
+	response, err := c.httpClient.Do(request)
+	if err != nil {
+		return nil, fmt.Errorf("do request: %v", err)
+	}
+	defer func() {
+		cErr := response.Body.Close()
+		if err == nil && cErr != nil {
+			err = fmt.Errorf("close response body: %v", cErr)
+		}
+	}()
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("read response body: %v", err)
+	}
+	fmt.Println("hoge response: ", response)
+
+	if err = errorFromResponseStatus(response, string(body)); err != nil {
+		return nil, err
+	}
+
+	var data GetAllUsersResponse
+	if err = json.Unmarshal(body, &data); err != nil {
+		return nil, fmt.Errorf("unmarshal response body: %v", err)
+	}
+
+	return data.Users, nil
+}
+
+func errorFromResponseStatus(r *http.Response, responseBody string) error {
+	// TODO(dfurman): return actionable errors
+	if r.StatusCode >= http.StatusBadRequest {
+		return fmt.Errorf("API response error, status: %v, response body: %v", r.Status, responseBody)
+	}
+	return nil
+}
+
+// payloads
+
+type GetAllUsersResponse struct {
+	Users []User `json:"users"`
+}
+
+type GetUserResponse struct {
+	User User `json:"user"`
+}
+
+// models
+
+// TODO(dfurman): consider splitting payloads and models
+
+type User struct {
+	Id           int    `json:"id"`
+	Username     string `json:"username"`
+	UserFullName string `json:"user_full_name"`
+	UserEmail    string `json:"user_email"`
+	Role         string `json:"role"`
+	EmailService bool   `json:"email_service"`
+	EmailProduct bool   `json:"email_product"`
+	LastLogin    string `json:"last_login"`
+	CreatedDate  string `json:"created_date"`
+	UpdatedDate  string `json:"updated_date"`
+	CompanyID    int    `json:"company_id"`
```

Comment:
how can we approach optional fields in payloads?

Reply:
This is what we need to resolve while implementing create/update methods. What comes to my mind is using fields of pointer types.